processing/src/main/java/io/druid/segment/data/GenericIndexed.java

```diff
@@ -147,31 +393,96 @@ public int indexOf(T value)
     return bufferIndexed.iterator();
   }
 
-  private final ByteBuffer theBuffer;
-  private final ObjectStrategy<T> strategy;
-  private final boolean allowReverseLookup;
-  private final int size;
+  public long getSerializedSize()
+  {
+    if (valueBuffers.size() != 1) {
+      throw new UnsupportedOperationException("Method not supported for version 2 GenericIndexed.");
+    }
+    return theBuffer.remaining() + 2 + 4 + 4; //2 Bytes for version and sorted flag. 4 bytes to store numbers
+                                              // of bytes and next 4 bytes to store number of elements.
+  }
 
-  private final int valuesOffset;
-  private final BufferIndexed bufferIndexed;
+  public void writeToChannel(WritableByteChannel channel) throws IOException
+  {
+    //version 2 will always have more than one buffer in valueBuffers.
+    if (valueBuffers.size() == 1) {
+      channel.write(ByteBuffer.wrap(new byte[]{VERSION_ONE, allowReverseLookup ? (byte) 0x1 : (byte) 0x0}));
+      channel.write(ByteBuffer.wrap(Ints.toByteArray(theBuffer.remaining() + 4))); // 4 Bytes to store size.
+      channel.write(ByteBuffer.wrap(Ints.toByteArray(size)));
+      channel.write(theBuffer.asReadOnlyBuffer());
+    } else {
+      throw new UnsupportedOperationException(
+          "GenericIndexed serialization for V2 is unsupported. Use GenericIndexedWriter instead.");
+    }
+  }
 
-  GenericIndexed(
-      ByteBuffer buffer,
-      ObjectStrategy<T> strategy,
-      boolean allowReverseLookup
-  )
+  /**
+   * Create a non-thread-safe Indexed, which may perform better than the underlying Indexed.
+   *
+   * @return a non-thread-safe Indexed
+   */
+  public GenericIndexed<T>.BufferIndexed singleThreaded()
   {
-    this.theBuffer = buffer;
-    this.strategy = strategy;
-    this.allowReverseLookup = allowReverseLookup;
+    if (valueBuffers.size() == 1) {
+      final ByteBuffer copyBuffer = valueBuffers.get(0).asReadOnlyBuffer();
+      return new BufferIndexed()
+      {
+        @Override
+        public T get(final int index)
+        {
+          checkIndex(index, size);
+
+          final int startOffset;
+          final int endOffset;
+
+          if (index == 0) {
+            startOffset = 4;
+            endOffset = headerBuffer.getInt(0);
+          } else {
+            int headerPosition = (index - 1) * Ints.BYTES;
+            startOffset = headerBuffer.getInt(headerPosition) + 4;
+            endOffset = headerBuffer.getInt(headerPosition + 4);
+          }
+          return _get(copyBuffer, startOffset, endOffset);
+        }
+      };
+    } else {
+
+      final List<ByteBuffer> copyValueBuffers = new ArrayList<>();
+      for (ByteBuffer buffer : valueBuffers) {
+        copyValueBuffers.add(buffer.asReadOnlyBuffer());
+      }
+
+      return new BufferIndexed()
```

Comment:
Both BufferIndexed are different as one at line 174 guarantees thread safety and one at line 456 does not.
BufferIndexed at line 174 won't be thread safe if we wrap `valueBuffers` with `Lists.transform(valueBuffers, b -> b.asReadOnlyBuffer())`, it creates a new object  
```final ByteBuffer copyBuffer = valueBuffers.get(fileNum).asReadOnlyBuffer();```
 and passes it to `_get()` method which sets the buffer position and passes it to object strategy. 
 However I'm reusing ByteBuffer objects in non thread safe version for get () operation.
```
      final List<ByteBuffer> copyValueBuffers = new ArrayList<>();
      for (ByteBuffer buffer : valueBuffers) {
        copyValueBuffers.add(buffer.asReadOnlyBuffer());
      }
```